controllers/application_controller.go

```diff
@@ -0,0 +1,247 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/retry"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	appv1beta1 "sigs.k8s.io/application/api/v1beta1"
+)
+
+const (
+	loggerCtxKey = "logger"
+)
+
+// ApplicationReconciler reconciles a Application object
+type ApplicationReconciler struct {
+	client.Client
+	Mapper meta.RESTMapper
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=app.k8s.io,resources=applications,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=app.k8s.io,resources=applications/status,verbs=get;update;patch
+
+func (r *ApplicationReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	rootCtx := context.Background()
+	logger := r.Log.WithValues("application", req.NamespacedName)
+	ctx := context.WithValue(rootCtx, loggerCtxKey, logger)
+
+	var app appv1beta1.Application
+	err := r.Get(ctx, req.NamespacedName, &app)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	// Application is in the process of being deleted, so no need to do anything.
+	if app.DeletionTimestamp != nil {
+		return ctrl.Result{}, nil
+	}
+
+	newApplicationStatus := r.getNewApplicationStatus(ctx, &app)
+
+	if equality.Semantic.DeepEqual(newApplicationStatus, &app.Status) {
+		return ctrl.Result{}, nil
+	}
+
+	newApplicationStatus.ObservedGeneration = app.Generation
```

Comment:
Should we do this before we check whether there are any changes to the status and possibly return early? Seems like the application has been observed at the given version even if there are no changes to the status.

Reply:
If we move this before we check whether there are changes to the status, the observedGeneration field will keep increasing. I'll take a deep look into why it is happening. Before we figure it out, I suggest we keep it as is to make sure we get correct observedGeneration field.